### PR TITLE
Beat Detection Tweak

### PR DESCRIPTION
**Pull request by OfficialIncubo**
I fixed the beat detection from hearing the treble (high frequencies) to bass, mid and treb variables.
Now I made the perfect and accurate beat detection for any music you want.

Check [this video demo](https://twitter.com/OctobellaMilkD/status/1625587045174476800?s=20) how I fixed the beat detection.

Plus, this can be reacted on both channels (mono), like projectM.

### CODE REVIEW

**Comment by milkdrop2077** [comment]
MilkDrop has been like this for the past 20 years, and changing the beat detection will make many presets react differently than intended.
I'm not sure what I prefer : https://www.youtube.com/watch?v=Ip6AsX1Tdao
It really depends on the kind of music you're listening and the kind of preset you're watching.

**Comment by milkdrop2077** [comment]
With the first minute of that song for example : https://www.youtube.com/watch?v=s_nc1IVoMxc
I prefer the old beat detection.

**Comment by OfficialIncubo** [comment]
If you prefer the old beat detection, i don't like it :(

If you hear some low frequencies, then boom, the red dot goes up!
For the high frequencies, you need some cymbals, hats, snakes and so on. The blue dot goes up if the treble (high frequencies) reacted.

Why don't you see the [20hz-20000hz frequency test](https://youtu.be/qNf9nzvnd1k) to show how it works?

**Comment by munibtahir** [comment]
I think backwards compatibility is a big concern if the beat detection is changed. There aren't many people actively authoring presets anymore, so we are essentially just stuck with the old presets.

**Comment by OfficialIncubo** [comment]
You see a little white dot that reacts in a different beat detections using my improved one.

https://user-images.githubusercontent.com/74858188/235320401-e872f439-0863-4184-af60-560590170005.MP4

I created the q1 variable in per-frame code and added the changeable beat detection mode with smoothing, such as bass_att, mid_att and treb_att. I recommend you to use headphones or subwoofers (or speakers with bass) before trying that.

In conclusion, my improved beat detection that I implemented gone too accurate than before. The old one reacts only high frequencies to bass, mid and treb values.

**Comment by Cmely** [comment]
First, thank you Milkdrop2077, for your work : I was looking for a easy to use, nice looking visualizer for my music, and it seems you've made the perfect combination :)

I'm not sure however the beat detection is worrking : would it be possible to use it to have it change only at the end of each songs ? And if 'yes', how would I achieve that ? And if 'no', do you think it could be possible / doable ?

**Comment by OfficialIncubo** [comment]
>...changing the beat detection will make many presets react differently than intended.

Well, the problem is not on presets using beat detection variables. It's the beat detection itself because it only jumps to the high frequencies reacting to bass, mid and treb variables.

**Comment by MrKlorox** [comment]
What if the new beat detection algorithm was just used for transitioning presets? Or if presets built with the new detection were flagged to use that algorithm, while the rest use the original? I think there's definitely use for better beat detection, though I get the compatibility issue.

**Comment by JosepMaJAZ** [comment]
backwards compatibility issue aside, I believe the implementation of this PR is wrong.


First, let's understand what we have in fSpecLeft/fSpecRight:
[https://github.com/milkdrop2077/MilkDrop3/blob/7569814634436b4470ebf25c0f9c13f435169b9e/code/vis_milk2/fft.cpp](fft.cpp)

First, it says that the range of that FFT is from 0 to 1/4th of the sampling rate.
So 0 = 0Hz and 576 = 1/4th or 11Khz for a 44Khz file.

It also says that the frequency linearly increases. With that, 11Khz/576 samples = 19.14Hz each output sample
It also suggests these 3 ranges for bass/mid/treble:
```
    //          1) 200 - 200*2^1.928                    or  200  - 761   Hz
    //          2) 200*2^1.928 - 200*2^(1.928*2)        or  761  - 2897  Hz
    //          3) 200*2^(1.928*2) - 200*2^(1.928*3)    or  2897 - 11025 Hz

```
This turns into samples like this:
761/19.41 =  39 ( 1 to 39 -1 )  (Note: I recommend avoiding sample zero. That one is the DC offset)
2897/19.41 = 149 ( 39 to 149 -1 )
And obviously, 11025 = 576 ( 149 to 576 -1)

The old implementation does:
0 to 1/6th ( 0-1837 Hz ) bass  (0 to 96 -1)
1/6th to 2/6th ( 1837-3675 Hz) mid (96 to 192 -1)
2/6th to 3/6th ( 3675-7350 Hz) high ( 192 to 288 - 1)

The proposed implementation in this PR does:
0 to 1/192th ( 0-57 Hz) bass  ( 0 to 3 - 1)
1/64th to 2/64th ( 172-344 Hz) mid ( 9 to 18 - 1)
2/6th to 3/6th ( 3675-7350 Hz) high ( 192 to 288 -1)

One can argue that the old bass range is too wide, but the proposed bass and mid range is way off and leaves holes between ranges.

So, In case anyone pretends to fix this, the fix should obviously be what the FFT algorithm comemnts suggest.

Then, it could be a user option to choose between old and new beat detector.

**Comment by OfficialIncubo** [comment]
Just looked at the loopback-capture's supported sample rate and it only supports 44100hz and 48000hz. Did you try to change any sample rate and try it with Milkdrop3 if it works?
If not, then I'll try myself.

**Comment by milkdrop2077** [comment]
Thank you @JosepMaJAZ for the technical details!
I have a question for you since you seem to know a lot about audio, I can't get MilkDrop3/BeatDrop to react to the microphone device with the loopback-capture. Any idea what could cause the problem? Many thanks

**Comment by damianofalcioni** [comment]
> What if the new beat detection algorithm was just used for transitioning presets? Or if presets built with the new detection were flagged to use that algorithm, while the rest use the original? I think there's definitely use for better beat detection, though I get the compatibility issue.

I think is easy to just add an option in the configuration to specify which algorithm to use
